modules/core/04-channel/keeper/export_test.go

```diff
@@ -0,0 +1,15 @@
+package keeper
+
+/*
+	This file is to allow for unexported functions to be accessible to the testing package.
```

Comment:
`ValidateUpgradeFields` is used by our tests

Reply:
I do wonder if there's a better way to do this in go. Is there no way to have intra-module public symbols? If not, this sort of wrapping with a good docstring is a +1 from me.

Reply:
can we move [this test code](https://github.com/cosmos/ibc-go/blob/04-channel-upgrades/modules/core/04-channel/keeper/keeper_test.go#L489) to upgrade.go and rename to `TestValidateUpgradeFields` ?

Reply:
I'm actually not a huge fan of this approach, in general I don't really like the pattern of defining methods on types outside the file the type is defined in, I find it makes it harder to reason about and I often find myself needing to rely on tools to determine what methods a type actually has. I think this is particularly tough as lots of types have the same name. E.g. `Keeper` so there is already some cognitive overhead to figure out what is going on. 

I know this isn't a new pattern being introduced in this PR, but I thought I would share my opinion on it here!

I do agree that it would be nicer if there was a standard way of having an unexported method be called from a test (without resorting to the `reflect` package )

I don't think it's the worst thing to have this method exported as is, but I do not feel super strongly about this, and am happy to go with this approach if the majority of the team likes it!

Reply:
> I do wonder if there's a better way to do this in go.

If you figure something out, I'd love to give it a go! From my knowledge there isn't a workaround, until I recently saw this construction I didn't realize there was a way you could unit test private functions

> but also happy to cover the code from the larger handler tests if needs be.

In this situation, it'd be totally fine to test in a larger handler, but in general I think we should be capable of unit testing private functions. If we manage to limit our exported API to msg server handler + grpc, we will end up with very large tests trying to test multiple layers. It's a lot more sustainable in my opinion to have unit tests which directly test a functions functionality. 

> I'm actually not a huge fan of this approach, in general I don't really like the pattern of defining methods on types outside the file the type is defined in, I find it makes it harder to reason about and I often find myself needing to rely on tools to determine what methods a type actually has. I think this is particularly tough as lots of types have the same name. E.g. Keeper so there is already some cognitive overhead to figure out what is going on.

I agree, but I do see a lot of value in being able to test unexported functions, in addition these public functions are defined in the testing file, so I don't think they should show up on go docs, as they shouldn't be exported outside of this directory. Happy to go with another approach, but I think it'd be good to move forward with the ability to test unexported functions for the time being. 

@chatton what if I rename the public function to `k.TestValidateUpgradeFields` or some prefix indicating that the function is for testing purposes only?

Reply:
Ah right. I remember I was reading about this yesterday and wanted to comment.

This is the typical approach for doing this, e.g [go's `fmt` does the same thing](https://cs.opensource.google/go/go/+/refs/tags/go1.20.4:src/fmt/export_test.go) with a couple of function. 

I guess you could also simplify by re-binding the names instead of re-declaring a public symbol but for methods this results in a call where the receiver is passed in explicitly as the first argument, i.e instead of:

```go
ValidateUpgradeFields(args)
```

that we have now it would be:

```go
// k is keeper
ValidateUpgradeFields(k, args)
```

not sure if this is something people really like (I personally don't) but its an alternative.